implementations/rust/vault/src/lib.rs

```diff
@@ -1,24 +1,96 @@
+    trivial_casts,
+    trivial_numeric_casts,
+    unconditional_recursion,
+    unused_import_braces,
+    unused_lifetimes,
+    unused_qualifications,
+    unused_extern_crates,
+    unused_parens,
+    while_true
+)]
+
 //! Implements the Ockam vault interface and provides
 //! a C FFI version.
 //!
 //! Vault represents a location where cryptographic keys live such
 //! as secure enclaves, TPMs, HSMs, Keyrings, files, memory, etc.
 
-#![deny(
-missing_docs,
-trivial_casts,
-trivial_numeric_casts,
-unconditional_recursion,
-unused_import_braces,
-unused_lifetimes,
-unused_qualifications,
-unused_extern_crates,
-unused_parens,
-while_true
-)]
+#![cfg_attr(feature = "nightly", feature(doc_cfg))]
+
+use crate::error::VaultFailError;
+use zeroize::Zeroize;
 
+/// Internal macros
+#[macro_use]
+mod macros;
 /// Represents the errors that occur within a vault
 pub mod error;
 #[cfg(feature = "ffi")]
 /// The ffi functions, structs, and constants
 pub mod ffi;
+/// The various enumerations of options
+pub mod types;
+
+use types::*;
+
+/// Represents the methods available to a Vault
+pub trait Vault: Zeroize {
+    /// Generate random bytes and fill them into `data`
+    fn random(&mut self, data: &mut [u8]) -> Result<(), VaultFailError>;
+    /// Compute the SHA-256 digest given input `data`
+    fn sha256<B: AsRef<[u8]>>(&self, data: B) -> Result<[u8; 32], VaultFailError>;
+    /// Create a new secret key
+    fn secret_generate(
+        &mut self,
+        attributes: SecretKeyAttributes,
+    ) -> Result<SecretKeyContext, VaultFailError>;
+    /// Import a secret key into the vault
+    fn secret_import(
+        &mut self,
+        secret: &SecretKey,
+        context: SecretKeyContext,
+        attributes: SecretKeyAttributes,
+    ) -> Result<(), VaultFailError>;
+    /// Export a secret key from the vault
+    fn secret_export(&self, context: SecretKeyContext) -> Result<SecretKey, VaultFailError>;
+    /// Set the attributes for a secret key
+    fn secret_attributes_get(
+        &self,
+        context: SecretKeyContext,
+    ) -> Result<SecretKeyAttributes, VaultFailError>;
+    /// Change the type for the secret key
+    fn secret_type_set(
+        &mut self,
+        context: SecretKeyContext,
+        xtype: SecretKeyType,
+    ) -> Result<(), VaultFailError>;
+    /// Remove a secret key from the vault
+    fn secret_destroy(&mut self, context: SecretKeyContext) -> Result<(), VaultFailError>;
+    /// Compute Elliptic-Curve Diffie-Hellman using this secret key
+    /// and the specified public key
+    fn ec_diffie_hellman(
+        &mut self,
+        context: SecretKeyContext,
+        peer_public_key: [u8; 32],
+    ) -> Result<SecretKey, VaultFailError>;
+    /// Compute the HKDF-SHA256 using the specified salt and input key material
+    /// and return the output key material of the specified length
+    fn hkdf_sha256<B: AsRef<[u8]>>(&mut self, salt: B, ikm: B, okm_len: usize) -> Result<Vec<u8>, VaultFailError>;
```

Comment:
Isn't it supposed to have separate template type for salt and for ikm?

Reply:
HKDF extract takes `salt` and `ikm` together and outputs the `prk`. 
HKDF expand takes optional `info` which is usually left blank with the `prk`

See [here](https://tools.ietf.org/html/rfc5869)

Reply:
I know how HKDF works, the question is about template <B: AsRef<[u8]>. As far as I understood it's to make calling the function more convenient e.g. you can call it with slice or with vector. But what if ikm is vector and salt is slice (which is very likely to happen because ikm is something you can get after ECDH and it can be Vector, while salt may be some static const represented as slice). Wouldn't this function fail in that case? Does it make sense to have template <B1: AsRef<[u8]>, <B2: AsRef<[u8]>> and salt: B1, ikm: B2?

Reply:
No because both of them implement the AsRef<[u8]> trait.

Reply:
@mikelodder7 I prepared an example to explain what I'm talking about https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=324e36f5bc058b23dc0f81c1c43126dc

Reply:
I haven't ever seen this problem in production. Perhaps we consider this change later. Thanks for the suggestion.